runtime/lua/vim/fs.lua

```diff
@@ -40,15 +42,34 @@ function M.dirname(file)
   if file == nil then
     return nil
   end
-  return vim.fn.fnamemodify(file, ':h')
+  vim.validate({ file = { file, 's' } })
+  if file:match('^%w:[\\/]?$') and iswin then
+    return file:gsub('\\', '/')
+  elseif not file:match('[\\/]') then
+    return '.'
+  elseif file:match('^/[^/]+$') or file == '/' then
```

Comment:
Similar comment here, it's cheaper/easier to check `file == '/'` so put that in the condition first so that we don't call `string.match()` if we don't need to.